source/ui/netsender.c

```diff
@@ -0,0 +1,528 @@
+}
+
+/*---------------------------------------------------------------------------------
+	Subtract the `struct timeval' values Y from X,
+	storing the result in RESULT.
+	Return 1 if the difference is negative, otherwise 0.
+
+	From http://www.gnu.org/software/libtool/manual/libc/Elapsed-Time.html
+---------------------------------------------------------------------------------*/
+static int timeval_subtract (struct timeval *result, struct timeval *x, struct timeval *y)
+{
+	struct timeval tmp;
+	tmp.tv_sec = y->tv_sec;
+	tmp.tv_usec = y->tv_usec;
+
+	// Perform the carry for the later subtraction by updating y.
+	if (x->tv_usec < tmp.tv_usec)
+	{
+		int nsec = (tmp.tv_usec - x->tv_usec) / 1000000 + 1;
+		tmp.tv_usec -= 1000000 * nsec;
+		tmp.tv_sec += nsec;
+	}
+
+	if (x->tv_usec - tmp.tv_usec > 1000000)
+	{
+		int nsec = (x->tv_usec - tmp.tv_usec) / 1000000;
+		tmp.tv_usec += 1000000 * nsec;
+		tmp.tv_sec -= nsec;
+	}
+
+	// Compute the time remaining to wait. tv_usec is certainly positive.
+	result->tv_sec = x->tv_sec - tmp.tv_sec;
+	result->tv_usec = x->tv_usec - tmp.tv_usec;
+
+	// Return 1 if result is negative.
+	return x->tv_sec < tmp.tv_sec;
+}
+
+static void timeval_add (struct timeval *result, struct timeval *x, struct timeval *y)
+{
+	result->tv_sec = x->tv_sec + y->tv_sec;
+	result->tv_usec = x->tv_usec + y->tv_usec;
+
+	if (result->tv_usec > 1000000)
+	{
+		result->tv_sec += result->tv_usec / 1000000;
+		result->tv_usec = result->tv_usec % 1000000;
+	}
+}
+
+static struct in_addr find3DS(int retries)
+{
+	struct sockaddr_in s, remote, rs;
+	char recvbuf[256];
+	char mess[] = "3dsboot";
+
+	int broadcastSock = socket(PF_INET, SOCK_DGRAM, 0);
+	if (broadcastSock < 0)
+	{
+		netsenderError("broadcast socket", errno);
+		remote.sin_addr.s_addr = INADDR_NONE;
+		return remote.sin_addr;
+	}
+
+	memset(&s, 0, sizeof(struct sockaddr_in));
+	s.sin_family = AF_INET;
+	s.sin_port = htons(NETWORK_PORT);
+	s.sin_addr.s_addr = INADDR_BROADCAST;
+
+	memset(&rs, 0, sizeof(struct sockaddr_in));
+	rs.sin_family = AF_INET;
+	rs.sin_port = htons(NETWORK_PORT);
+	rs.sin_addr.s_addr = INADDR_ANY;
+
+	int recvSock = socket(PF_INET, SOCK_DGRAM, 0);
+
+	if (recvSock < 0)
+	{
+		netsenderError("receive socket", errno);
+		close(broadcastSock);
+		remote.sin_addr.s_addr = INADDR_NONE;
+		return remote.sin_addr;
+	}
+
+	if (bind(recvSock, (struct sockaddr*) &rs, sizeof(rs)) < 0)
+	{
+		netsenderError("bind receive socket", errno);
+		close(broadcastSock);
+		close(recvSock);
+		remote.sin_addr.s_addr = INADDR_NONE;
+		return remote.sin_addr;
+	}
+
+	fcntl(recvSock, F_SETFL, O_NONBLOCK);
+	struct timeval wanted, now, result;
+
+	gettimeofday(&wanted, NULL);
+
+	int timeout = retries, len;
+	while (timeout)
+	{
+		if (wantExit)
+			break;
+
+		gettimeofday(&now, NULL);
+		if (timeval_subtract(&result,&wanted,&now))
+		{
+			if (sendto(broadcastSock, mess, strlen(mess), 0, (struct sockaddr *)&s, sizeof(s)) < 0)
+			{
+				netsenderError("sendto", errno);
```

Comment:
I tested, and apparently this error breaks the UI logic, because uiExitState ends up being called twice (one here, and the other in netsenderTask)